Fix #1233, Add properties missing from User and Activity models

Chris-Johnston opened the pull request:
Fix #1233 

This PR adds some properties that were missing from the API models, and implements them in the Socket and Rest types.

**Activity:**
- Adds the Activity `Flags` property. This determines which actions a user can take (join a game, watch a stream, listen along)
   - Added the ActivityFlag enum for each of these flags

**SelfUser:**
- Adds the SelfUser `Flags` property. This determines some properties about a user (hypesquad, early nitro adopter, staff)
  - Adds the `UserProperties` type. Some of these flags are not in the official documentation, and were determined with developer tools.
- Adds the `PremiumType` property and enum, which is the subscription that is active for a user's account.
- Adds the `Locale` property, which is the user's selected locale in IETF (`en-US`, `zh-TW`, etc.) format.

Chris-Johnston committed:
WIP commit, update IActivity and IUser with missing types

Chris-Johnston committed:
incomplete implementation of interfaces

Chris-Johnston committed:
Add implementation of activity flags, and extension method to check flag

Chris-Johnston committed:
remove usings that were not required

Chris-Johnston committed:
clean up files and add documentation where missing

Chris-Johnston committed:
remove unused usings

Chris-Johnston committed:
Add remark saying that premium_type info may be inaccessible to bots

it seems that this information can not be accessed by any user,
and may require the identity oauth scope

Chris-Johnston committed:
Add locale property to the user model

adds the locale property to the user model, which appears not to be
specified for bots

Chris-Johnston committed:
rename ActivityFlags

Chris-Johnston committed:
Add incomplete xmldoc to ActivityFlag

Added xmldoc to the ActivityFlag type, excluded flags that i was unsure about

Chris-Johnston committed:
rename the UserFlags file

Chris-Johnston committed:
Add incomplete xmldoc to UserFlag enum

Chris-Johnston committed:
Remove unnecessary extension methods and use built-in functionality

Joe4evr commented on src/Discord.Net.Core/Entities/Activities/ActivityFlag.cs:
```diff
@@ -0,0 +1,31 @@
+    ///     These describe what the activity payload includes.
+    /// </summary>
+    [Flags]
+    public enum ActivityFlag
```
A flags enum is [recommended to be named in plural and without `Flag`/`Flags` suffix](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces#naming-enumerations).

Chris-Johnston replied:
[I was unsure if we should stray from the name used by the Discord API docs](https://discordapp.com/developers/docs/topics/gateway#activity-object-activity-flags). I guess `ActivityProperties` and `UserProperties` might be alright alternatives?

Joe4evr commented on src/Discord.Net.Core/Entities/Users/UserFlag.cs:
```diff
@@ -0,0 +1,42 @@
+namespace Discord
+{
+    [Flags]
+    public enum UserFlag
```
Ditto on the enum naming.

Joe4evr commented on src/Discord.Net.Core/Entities/Users/IUser.cs:
```diff
@@ -112,5 +112,33 @@ public interface IUser : ISnowflakeEntity, IMentionable, IPresence
+        ///     meaning that users and bots will not have access to this information.
+        /// </remarks>
+        /// <returns>
+        ///     The type of Nitro subscription the user subscribes to, or null if this value could not be determined.
```
@Still34 Don't we always want language keywords (like `null`) highlighted in the docs? Need to know if this needs changing for consistency.

Chris-Johnston replied:
iirc for language keywords like `null` we use `<c>null</c>`

Still34 replied:
Yes, we do. We use <c> tag to wrap out null.

Joe4evr commented on src/Discord.Net.Core/Entities/Activities/ActivityFlag.cs:
```diff
@@ -0,0 +1,31 @@
+        ///     Indicates that this activity can be spectated.
+        /// </summary>
+        Spectate = 0b100,
+        JoinRequest = 0b1000,
```
Missing doc comment.

Chris-Johnston replied:
I was a bit unsure about when this flag is used, I think it's used whenever there's an "Ask to join" button on an activity?

Joe4evr commented on src/Discord.Net.Core/Entities/Users/PremiumType.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public enum PremiumType
+    {
+        None = 0,
```
Missing doc comment.

Joe4evr commented on src/Discord.Net.Core/Entities/Users/UserFlag.cs:
```diff
@@ -0,0 +1,42 @@
+        ///     Flag given to Discord partners.
+        /// </summary>
+        Partner = 0b10,
+        HypeSquadEvents = 0b100,
```
Missing doc comment.

Chris-Johnston replied:
Here also I was unsure about where this is used. I think this may have been the old hypesquad flag, before they introduced the houses?

Chris-Johnston replied:
I removed this flag in 802d489f474e5e0b0f4e0a9cc95cdba354df4363 , as it seems that this flag has been replaced with the new HypeSquad flags.

Joe4evr commented on src/Discord.Net.Core/Entities/Activities/ActivityFlag.cs:
```diff
@@ -0,0 +1,31 @@
+    [Flags]
+    public enum ActivityFlag
+    {
+        Instance = 1,
```
Missing doc comment and a 0 value.

Chris-Johnston replied:
Added a 0 value. Wasn't sure what this specific value is used for, so I'm not sure what to write in xmldocs.

Joe4evr reviewed (commented):
This might be overly nit-picky, but I'm practicing doing code review, as I suspect I'm gonna do a lot of it in the near future.

Chris-Johnston commented:
Not nit-picky at all, I appreciate the detailed review.

Chris-Johnston committed:
use <c> tag for xmldoc 'null's

Chris-Johnston committed:
Add xmldoc for JoinRequest ActivityFlag

Chris-Johnston committed:
improve the PremiumType xmldoc

Chris-Johnston committed:
Add zero activityflag

Still34 commented on src/Discord.Net.Core/Entities/Activities/IActivity.cs:
```diff
@@ -19,5 +19,15 @@ public interface IActivity
         /// </returns>
         ActivityType Type { get; }
+        /// <summary>
+        ///     The flags that are relevant to this activity.
```
Verb is required, i.e. `Gets....`

Still34 commented on src/Discord.Net.Core/Entities/Users/IUser.cs:
```diff
@@ -112,5 +112,33 @@ public interface IUser : ISnowflakeEntity, IMentionable, IPresence
         /// </returns>
         Task<IDMChannel> GetOrCreateDMChannelAsync(RequestOptions options = null);
+        /// <summary>
+        ///     The flags that are applied to a user's account.
```
Ditto with summary verb.

Still34 commented on src/Discord.Net.Core/Entities/Users/IUser.cs:
```diff
@@ -112,5 +112,33 @@ public interface IUser : ISnowflakeEntity, IMentionable, IPresence
+        /// </returns>
+        UserFlag Flags { get; }
+        /// <summary>
+        ///     The type of Nitro subscription that is active on this user's account.
```
Ditto with summary verb.

Still34 commented on src/Discord.Net.Core/Entities/Users/IUser.cs:
```diff
@@ -112,5 +112,33 @@ public interface IUser : ISnowflakeEntity, IMentionable, IPresence
+        /// </returns>
+        PremiumType? PremiumType { get; }
+        /// <summary>
+        ///     The user's chosen language option.
```
Ditto with summary verb.

Still34 commented on src/Discord.Net.Core/Entities/Users/IUser.cs:
```diff
@@ -112,5 +112,33 @@ public interface IUser : ISnowflakeEntity, IMentionable, IPresence
+        /// </summary>
+        /// <returns>
+        ///     The value of the user's chosen language option, if provided.
+        /// </returns>
```
An example of the returned value would be preferred, since a `string` type is rather generic.

Chris-Johnston committed:
Add summary tag verbs, example of IUser locale

Chris-Johnston committed:
Rename Flag enum types to use Properties suffix

Still34 commented:
It doesn't look like the `Details` field is being exposed into the public `Game` model?

Chris-Johnston commented:
Odd, so after some looking `Details` is currently used to populate the `TrackTitle` property of `SpotifyGame`.

Chris-Johnston committed:
Expose the details field in the Game model

Chris-Johnston committed:
update Activity extension methods to support Details field

Chris-Johnston committed:
Use inherited Details implementation in RichGame

Chris-Johnston committed:
lint: remove commented out code from debugging

Chris-Johnston committed:
Fix issue in previous commit

Chris-Johnston committed:
Move oauth-only user fields to SelfUser classes

Moves the implementation of the Flags, PremiumType, and Locale user fields to ISelfUser classes. In testing, it seemed that normal bot accounts did not have this information supplied to them. When tested with a Bearer token in the Rest client, these fields are set.

Chris-Johnston committed:
remove old HypeSquadEvents flag that appears to have been replaced with the new HypeSquad flags
